pkg/networkserver/networkserver_util_test.go

```diff
@@ -558,50 +558,51 @@ var _ DeviceRegistry = MockDeviceRegistry{}
+	GetByEUIFunc    func(ctx context.Context, joinEUI, devEUI types.EUI64, paths []string) (*ttnpb.EndDevice, context.Context, error)
+	GetByIDFunc     func(ctx context.Context, appID ttnpb.ApplicationIdentifiers, devID string, paths []string) (*ttnpb.EndDevice, context.Context, error)
+	RangeByAddrFunc func(ctx context.Context, devAddr types.DevAddr, paths []string, f func(context.Context, *ttnpb.EndDevice) bool) error
+	SetByIDFunc     func(ctx context.Context, appID ttnpb.ApplicationIdentifiers, devID string, paths []string, f func(context.Context, *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error)) (*ttnpb.EndDevice, context.Context, error)
 }
 
 // GetByEUI calls GetByEUIFunc if set and panics otherwise.
-func (m MockDeviceRegistry) GetByEUI(ctx context.Context, joinEUI, devEUI types.EUI64, paths []string) (*ttnpb.EndDevice, error) {
+func (m MockDeviceRegistry) GetByEUI(ctx context.Context, joinEUI, devEUI types.EUI64, paths []string) (*ttnpb.EndDevice, context.Context, error) {
 	if m.GetByEUIFunc == nil {
 		panic("GetByEUI called, but not set")
 	}
 	return m.GetByEUIFunc(ctx, joinEUI, devEUI, paths)
 }
 
 // GetByID calls GetByIDFunc if set and panics otherwise.
-func (m MockDeviceRegistry) GetByID(ctx context.Context, appID ttnpb.ApplicationIdentifiers, devID string, paths []string) (*ttnpb.EndDevice, error) {
+func (m MockDeviceRegistry) GetByID(ctx context.Context, appID ttnpb.ApplicationIdentifiers, devID string, paths []string) (*ttnpb.EndDevice, context.Context, error) {
 	if m.GetByIDFunc == nil {
 		panic("GetByID called, but not set")
 	}
 	return m.GetByIDFunc(ctx, appID, devID, paths)
 }
 
 // RangeByAddr calls RangeByAddrFunc if set and panics otherwise.
-func (m MockDeviceRegistry) RangeByAddr(ctx context.Context, devAddr types.DevAddr, paths []string, f func(*ttnpb.EndDevice) bool) error {
+func (m MockDeviceRegistry) RangeByAddr(ctx context.Context, devAddr types.DevAddr, paths []string, f func(context.Context, *ttnpb.EndDevice) bool) error {
 	if m.RangeByAddrFunc == nil {
 		panic("RangeByAddr called, but not set")
 	}
 	return m.RangeByAddrFunc(ctx, devAddr, paths, f)
 }
 
 // SetByID calls SetByIDFunc if set and panics otherwise.
-func (m MockDeviceRegistry) SetByID(ctx context.Context, appID ttnpb.ApplicationIdentifiers, devID string, paths []string, f func(*ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error)) (*ttnpb.EndDevice, error) {
+func (m MockDeviceRegistry) SetByID(ctx context.Context, appID ttnpb.ApplicationIdentifiers, devID string, paths []string, f func(context.Context, *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error)) (*ttnpb.EndDevice, context.Context, error) {
 	if m.SetByIDFunc == nil {
 		panic("SetByID called, but not set")
 	}
 	return m.SetByIDFunc(ctx, appID, devID, paths, f)
 }
 
-type deviceAndError struct {
-	Device *ttnpb.EndDevice
-	Error  error
+type deviceContextAndError struct {
```

Comment:
nit; `contextualDeviceAndError` (like https://github.com/TheThingsNetwork/lorawan-stack/blob/master/pkg/applicationserver/io/io.go#L49, or `contextualDeviceError`; in Go you can concatenate functions like https://golang.org/pkg/io/#ReadWriteSeeker)